src/common/proxy_wasm/ngx_proxy_wasm_host.c

```diff
@@ -1575,7 +1576,142 @@ ngx_proxy_wasm_hfuncs_dequeue_shared_queue(ngx_wavm_instance_t *instance,
+                                          NGX_WAVM_ERROR);
+    }
+
+    prefixed_name.data = buf;
+    prefixed_name.len = ngx_sprintf(buf, "pw.%V.%V", filter_name, &name) - buf;
+
+    if (ngx_wa_metrics_add(metrics, &prefixed_name, type, id) != NGX_OK) {
+        return ngx_proxy_wasm_result_trap(pwexec, "could not define metric",
+                                          rets,
+                                          NGX_WAVM_ERROR);
+    }
+
+    return ngx_proxy_wasm_result_ok(rets);
+}
+
+
+static ngx_int_t
+ngx_proxy_wasm_hfuncs_increment_metric(ngx_wavm_instance_t *instance,
+    wasm_val_t args[], wasm_val_t rets[])
+{
+    uint32_t                metric_id;
+    ngx_int_t               rc, offset;
+    ngx_cycle_t            *cycle = (ngx_cycle_t *) ngx_cycle;
+    ngx_wa_metrics_t       *metrics = ngx_wasmx_metrics(cycle);
+    ngx_proxy_wasm_exec_t  *pwexec = ngx_proxy_wasm_instance2pwexec(instance);
+
+    metric_id = args[0].of.i32;
+    offset = args[1].of.i64;
+
+    rc = ngx_wa_metrics_increment(metrics, metric_id, offset);
+
+    if (rc != NGX_OK) {
+        return ngx_proxy_wasm_result_trap(pwexec, "could not increment metric",
+                                          rets, NGX_WAVM_ERROR);
+    }
+
+    return ngx_proxy_wasm_result_ok(rets);
+}
+
+
+static ngx_int_t
+ngx_proxy_wasm_hfuncs_record_metric(ngx_wavm_instance_t *instance,
+    wasm_val_t args[], wasm_val_t rets[])
+{
+    uint32_t                metric_id;
+    ngx_int_t               rc, offset;
+    ngx_cycle_t            *cycle = (ngx_cycle_t *) ngx_cycle;
+    ngx_wa_metrics_t       *metrics = ngx_wasmx_metrics(cycle);
+    ngx_proxy_wasm_exec_t  *pwexec = ngx_proxy_wasm_instance2pwexec(instance);
+
+    metric_id = args[0].of.i32;
+    offset = args[1].of.i64;
+
+    rc = ngx_wa_metrics_record(metrics, metric_id, offset);
+
+    if (rc != NGX_OK) {
+        return ngx_proxy_wasm_result_trap(pwexec,
+                                          "could not record metric value", rets,
+                                          NGX_WAVM_ERROR);
+    }
+
+    return ngx_proxy_wasm_result_ok(rets);
+}
+
+
+static ngx_int_t
+ngx_proxy_wasm_hfuncs_get_metric(ngx_wavm_instance_t *instance,
+    wasm_val_t args[], wasm_val_t rets[])
+{
+    uint32_t                metric_id;
+    ngx_int_t               rc;
+    ngx_uint_t             *ret_value;
+    ngx_cycle_t            *cycle = (ngx_cycle_t *) ngx_cycle;
+    ngx_wa_metrics_t       *metrics = ngx_wasmx_metrics(cycle);
+    ngx_proxy_wasm_exec_t  *pwexec = ngx_proxy_wasm_instance2pwexec(instance);
+
+    metric_id = args[0].of.i32;
+    ret_value = NGX_WAVM_HOST_LIFT(instance, args[1].of.i32, ngx_uint_t);
+
+    rc = ngx_wa_metrics_get(metrics, metric_id, ret_value);
+
+    if (rc != NGX_OK) {
+        /* TODO: format with key */
+        return ngx_proxy_wasm_result_trap(pwexec, "could not retrieve metric",
+                                          rets, NGX_WAVM_ERROR);
```

Comment:
So in the case when a metric does not exist, we have an error log from `ngx_wa_metrics_get` plus this trap? The below tests did not show anything about the trap.

How about we combine them? This unifies this error path's logs and for the future, the `NGX_DECLINED` status code will make for a more precise trap message for some other code paths if we need to distinguish failure cases.

Reply:
Also agreed that formatting the host traps with the key for all failure cases would be nice for the user - when the key is too long the nginx error log facility will already truncate the message for us if we write it smart.

Reply:
b9e981aaca0c9f2b81535d2c4e5b184131f55e6b